Major gates.rst revision

levbishop opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.

Are you changing the specification? This PR needs to be approved by the TSC members.

Are you changing the grammar to adjust to the specification?

The specification is the source of truth and any grammar updates should also coincide with accepted specification updates in the same PR.

-->

### Summary



### Details and comments

mhillenbrand commented on source/language/gates.rst:
```diff
@@ -5,172 +5,103 @@
-.. multifigure::
-   :rowitems: 2
+As well as gates that each represent a fixed unitary, OpenQASM also supports gates that represent *families* of unitaries, parameterized
+by angle variables.  Todistinguish the (optional) angle parameters from the (required) quantum arguments, if any angle parameters are
```
minor typo
```suggestion
by angle variables. To distinguish the (optional) angle parameters from the (required) quantum arguments, if any angle parameters are
```

mhillenbrand commented on source/language/gates.rst:
```diff
@@ -5,172 +5,103 @@
+A minimal compiler implementation might simply expand ``gate`` definitions repeatedly until reaching
+definitions for which :ref:`defcal blocks <pulse-gates>` are known. A more sophisticated implementation
+might use the `gate` definitions of the gates with associated ``defcal`` blocks to
+to build a gate library, and use methods based on KAK decompositions to rewrite into this hardware library.
```
nit: superfluous `to`
```suggestion
build a gate library, and use methods based on KAK decompositions to rewrite into this hardware library.
```

blakejohnson commented on source/language/gates.rst:
```diff
@@ -5,172 +5,103 @@
-   0 & 0 & 0 & 1 \\
-   0 & 0 & 1 & 0 \\
-   0 & 1 & 0 & 0 \end{array}\right)
+//FIXME: Is that what we really want? Maybe we could use attributes to mark which arguments are safe (commuting) to broadcast over
+and error in the case of non-commuting broadcasting?
```
I think this is what we agreed to, at least.

blakejohnson commented on source/language/gates.rst:
```diff
@@ -202,15 +133,19 @@ In general, new gates are defined by statements of the form
+arguments. The parameters are identifiers that behave as ``angle`` type with unknown
+size. A compiler might recognize certain constructs and replace them with mathematically-
+equivalent versions that would be true for arbitrary precision, or it might do calculations
+at a fixed angle size, for example corresponding to the size of angle parameters to the corresponding
```
```suggestion
at a fixed ``angle`` size, for example corresponding to the size of angle parameters to the corresponding
```

blakejohnson commented on source/language/gates.rst:
```diff
@@ -240,58 +175,32 @@ and these symbols are scoped only to the subroutine body.
 bit and :math:`C_U` is the controlled-:math:`U` gate. The new quantum argument is prepended to the
 argument list for the controlled-:math:`U` gate. The quantum argument can be a register, and in this
-case controlled gate broadcast over it (as it shown in examples above for CNOT gate). The modified
+case controlled gate broadcast over it (as for all gates //FIXME). The modified
```
What is the plan here?

levbishop replied:
The plan is that all gates broadcast with the same semantics

blakejohnson commented on source/language/gates.rst:
```diff
@@ -378,7 +293,73 @@ repetitions of ``inv @ U`` for :math:`k < 0`.
+
+For *non-parameterized gates*, the choice of ``U`` and ``gphase`` as the built-in gates, along with one
+two-qubit entangling gate CNOT as defined gives a universal gate set that can represent general n-qubit
+unitary with an :math:`O(2^n)` size description :cite:`barenco95`. This basis is not an enforced compilation
```
```suggestion
unitaries with an :math:`O(2^n)` size description :cite:`barenco95`. This basis is not an enforced compilation
```

levbishop commented:
I think I've addressed all remaining issues.
Possible controversial points:
1) I dropped the requirement that user-defined gates must have at least one qubit. It feels more consistent to treat gphase as a 0-qubit gate, and to allow user-defined 0-qubit gates eg: `gate negate {gphase(pi);}`
2) I kept the global phase of the standard gates. This maintains the backwards compatibility, but at the cost of having a few gates which are non-2pi-periodic: rx(), ry(), rz() (and by extension crx(), cry(), crz()), cu(), u2(), u3(). If we are already sucking up the breaking change of U() global phase maybe we should just bite the bullet and do the same for these gates too?

blakejohnson commented on source/language/gates.rst:
```diff
@@ -202,18 +135,21 @@ In general, new gates are defined by statements of the form
+arguments. The parameters are identifiers that behave as ``angle`` type with unknown
+size. A compiler might recognize certain constructs and replace them with mathematically-
+equivalent versions that would be true for arbitrary precision, or it might do calculations
+at a fixed ``angle`` size, for example corresponding to the size of ``angle`` parameters to the corresponding
```
```suggestion
at a fixed ``angle`` size, for example corresponding to the size of ``angle`` parameters in the corresponding
```

blakejohnson commented on source/language/gates.rst:
```diff
@@ -378,7 +294,87 @@ repetitions of ``inv @ U`` for :math:`k < 0`.
+rotating frame and the rotating wave approximation (RWA) holds. This is the domain covered by the OpenPulse
+specification. For this case, the only supported form of run-time parameterization
+will likely be via a ``rz(ϕ)`` implemented by specialized frame-tracking hardware.
+This gate is covered by the built-in ``U`` as a special case ``U(0, ϕ, θ)``
```
Do you mean to have zeros for the first two arguments of `U`?

levbishop replied:
I didn't but agree that would be clearer

blakejohnson reviewed (commented):
I've double checked all the expressions in `stdgates.inc` and it looks correct to me. A have just a couple small questions I would like resolved.

levbishop commented on source/language/gates.rst:
```diff
@@ -378,7 +294,87 @@ repetitions of ``inv @ U`` for :math:`k < 0`.
+rotating frame and the rotating wave approximation (RWA) holds. This is the domain covered by the OpenPulse
+specification. For this case, the only supported form of run-time parameterization
+will likely be via a ``rz(ϕ)`` implemented by specialized frame-tracking hardware.
+This gate is covered by the built-in ``U`` as a special case ``U(0, ϕ, θ)``
```
```suggestion
This gate is covered by the built-in ``U`` as a special case ``U(0, 0, ϕ)``
```

blakejohnson commented:
Thanks, Lev. I'm glad to get this merged!
